File: LFAutomation/Python/sequential_gating.py
# Import the .NET class library
import clr

# Import python sys module
import sys

# Import os module
import os

# Import System.IO for saving and opening files
from System.IO import *

# Import C compatible List and String
from System import String
from System.Collections.Generic import List

# Add needed dll references
sys.path.append(os.environ['LIGHTFIELD_ROOT'])
sys.path.append(os.environ['LIGHTFIELD_ROOT']+"\\AddInViews")
clr.AddReference('PrincetonInstruments.LightFieldViewV5')
clr.AddReference('PrincetonInstruments.LightField.AutomationV5')
clr.AddReference('PrincetonInstruments.LightFieldAddInSupportServices')

# PI imports
from PrincetonInstruments.LightField.Automation import Automation
from PrincetonInstruments.LightField.AddIns import CameraSettings
from PrincetonInstruments.LightField.AddIns import ExperimentSettings
from PrincetonInstruments.LightField.AddIns import GatingMode
from PrincetonInstruments.LightField.AddIns import Pulse, DeviceType



def set_sequential_gating(starting_width, starting_delay,
                   ending_width, ending_delay):
    # Check Gating Mode existence
    if (experiment.Exists(CameraSettings.GatingMode)):

        # Set sequential gating mode
        experiment.SetValue(CameraSettings.GatingMode, 
                            GatingMode.Sequential)   
        
        pulser = []

        # Add PI Pulse type with parameters to pulser list
        pulser.append(Pulse(starting_width, starting_delay))

        # Add PI Pulse type with parameters to pulser list
        pulser.append(Pulse(ending_width,ending_delay))      

        # Set sequential starting gate
        experiment.SetValue(
            CameraSettings.GatingSequentialStartingGate,
            pulser[0])

        # Set sequential ending gate
        experiment.SetValue(
            CameraSettings.GatingSequentialEndingGate,
            pulser[1])
    else:
        print("System not capable of Gating Mode")
        

def set_on_chip_accumulations(accumulations):
    # Set On-chip accumulations
    experiment.SetValue(
    CameraSettings.ReadoutControlAccumulations, accumulations)
    

def get_on_chip_accumulations():
    print(String.Format("{0} {1}", "Current On-Chip Accumulations:",
                        experiment.GetValue(
                            CameraSettings.ReadoutControlAccumulations)))
    
def device_found():
    # Find connected device
    for device in experiment.ExperimentDevices:
        if (device.Type == DeviceType.Camera and
            "PI-MAX" in device.Model):
            return True
     
    # If connected device is not a camera inform the user
    print("Camera not found. Please add ", 
          "PI-MAX type camera to LightField.")
    return False  

    
        
# Create the LightField Application (true for visible)
# The 2nd parameter forces LF to load with no experiment 
auto = Automation(True, List[String]())

# Get experiment object
experiment = auto.LightFieldApplication.Experiment

# If PI-MAX3 or 4 found continue
if (device_found()==True):

    # Set on-chip accumulations
    set_on_chip_accumulations(3)

    # Print on-chip accumulations
    get_on_chip_accumulations()

    # Set sequential starting and ending
    # widths and delays
    set_sequential_gating(100, 50, 1000, 50)

    # Set number of frames 
    experiment.SetValue(ExperimentSettings.AcquisitionFramesToStore, 10)

    # Acquire image
    experiment.Acquire()



#Result: This sample will set/get a value for on chip accumulations.
#        Set starting and ending sequential pulse width and delay.
#        Set number of frames.
#        Acquire an image.

File: LFAutomation/MATLAB/sequential_gating.m
clear all

Setup_LightField_Environment;
instance1=lfm(true);
exp=instance1.application.Experiment;

if pm_loaded(exp)
    instance1.set(PrincetonInstruments.LightField.AddIns.CameraSettings...
        .ReadoutControlAccumulations, 3);
    fprintf('Current On-Chip Accumulations: %8g\n',string(instance1...
        .get(PrincetonInstruments.LightField.AddIns.CameraSettings...
        .ReadoutControlAccumulations)));
    set_sequential_gate(exp, 100, 50, 1000, 50); %pulse values in ns
    set(instance1,PrincetonInstruments.LightField.AddIns...
        .ExperimentSettings.AcquisitionFramesToStore, 10)
    data = acquire(instance1);
end

instance1.close;

function camLoaded=pm_loaded(obj)
    
    camLoaded=0;
    devList=obj.ExperimentDevices;
    objects=devList.Count;

    for i=1:objects
        device=devList.Item(i-1);
        type=string(device.Type);
        model=string(device.Model);
        if strcmp(type,'Camera') && contains(model,'PI-MAX')
            camLoaded=1;
            break;
        end
    end
    if ~camLoaded
        fprintf('Compatible camera not found. Please add PI-MAX type camera to LightField.');
    end
end

function set_sequential_gate(obj,start_width,start_delay...
    ,end_width,end_delay)
    if obj.Exists(PrincetonInstruments.LightField.AddIns.CameraSettings...
            .GatingMode)
        obj.SetValue(PrincetonInstruments.LightField.AddIns.CameraSettings...
            .GatingMode, PrincetonInstruments.LightField.AddIns...
            .GatingMode.Sequential);
        obj.SetValue(PrincetonInstruments.LightField.AddIns.CameraSettings...
            .GatingSequentialStartingGate, PrincetonInstruments.LightField...
        .AddIns.Pulse(start_width,start_delay));
        obj.SetValue(PrincetonInstruments.LightField.AddIns.CameraSettings...
        .GatingSequentialEndingGate, PrincetonInstruments.LightField...
        .AddIns.Pulse(end_width,end_delay));
    else
        fprintf('System not capable of Gating Mode');
    end
end